rfcs/react-components/convergence/custom-styling.md

```diff
@@ -0,0 +1,475 @@
+  const state = useButton_unstable(props, ref);
+  const rootStyle = useCustomButtonStyle();
+  state.root.className = rootStyle;
+  useButtonStyles_unstable(state);
+
+  return renderButton_unstable(state);
+  // Casting is required due to lack of distributive union to support unions on @types/react
+}) as ForwardRefComponent<ButtonProps>;
+```
+
+### 💡 Improvement - Explicit APIs for modifying commonly modified styles
+
+checkboxCheckedBackground = 'foo' vs ':enabled:not(:checked):not(:indeterminate)': { background: 'foo'}
+
+### 💡 Improvement - Perf - complex selectors are much slower than local CSS variables
+
+Single className adding multiple css variables is much more performant than complex selectors targeting individual parts of the control
+
+```
+.foo {
+  --a: green;
+  --b: yellow;
+}
+
+vs
+
+.foo:enabled:not(:checked):not(:indeterminate)' > svg
+```
+
+### 👍
+
+### 👎
+
+Composition approach would be difficult to adjust at runtime because css vars are already set on the root, and new styles would either need to replace the old one, be more specific, or not use makeResetStyles so user could override individual css vars
+
+### 🤔
+
+## Option D: Build time component tokens
+
+Each component would define a hierarchy component-level token constants.
+Component tokens would reference alias tokens.
+The hierarchy would have levels for slot and state variations.
+The component style hook would then use the component token constants rather than alias token constants.
+The component's tokens could be substituted at build time for a mapping to other tokens (CSS var references)
+
+```ts
+// Button.tokens.ts (limited example)
+export const buttonTokens = {
+  root: {
+    backgroundColor: tokens.colorNeutralBackground1,
+    hover: {
+      backgroundColor: tokens.colorNeutralBackground1Hover,
+    }
+    active: {
+      backgroundColor: tokens.colorNeutralBackground1Pressed
+    },
+    outline: {
+      backgroundColor: tokens.colorTransparentBackground,
+      hover: {
+        backgroundColor: tokens.colorTransparentBackgroundHover
+      },
+      active: {
+        backgroundColor: tokens.colorTransparentBackgroundPressed
+      }
+    }
+  },
+  icon: {
+    small: {
+      spacing: tokens.spacingHorizontalXS
+    },
+    medium: {
+      spacing: tokens.spacingHorizontalSNudge
+    },
+    large: {
+      spacing: tokens.spacingHorizontalSNudge
+    }
+  }
+}
+```
+
+### 👍
+
+- Codifies the mapping of component part style to alias token.
+- Build time substitution allows for continued build time style optimizations
+- Runtime CSS vars are unchanged and works with existing themes
+- Provides a similar experience to custom styles and theme component styles from v8, but mapped to slots.
+
+### 👎
+
+- Exports yet another set of constants for the bundle
+- Adds some complexity to the style hooks
+
+### 🤔
+
+- Would the hierarchy of component tokens be a contract?
+
+## Decision making
+
+//TODO @Geoff
+
+## Appendix: Partner scenarios
```

Comment:
Do we have any other write up of the scenarios we want to address?
Why I am asking is that I generally see two different approaches in the RFC:
- Option A and B allow to override/extend the component with any arbitrary styles.
- Option C and D add 'component variables' as another extension option to the components.

The arbitrary styles allow the partners to tweak "everything" but also to break "anything", making the component styles part of the API.

On the other hand, component variables is an enumerated finite list of options which can be implemented without leaking the component styles. With that approach we can keep the implementation under control and covered by tests. Those might be complex and insufficient, though - we will not test all permutations of all variables and possible values.
Also, obviously, there will always be a `style` which will not be exposed as a component variable but a partner will urgently need to tweak that one - in that case, will we support both?

Reply:
I like option A and B :-) It's not our responsibility if teams break or deviate. It also saves us wasted time bickering about what should and shouldn't be exposed for customization.